packages/flutter_tools/lib/src/custom_devices/custom_device.dart

```diff
@@ -0,0 +1,736 @@
+    // copy the list so we don't modify it concurrently
+    List<ForwardedPort>.of(_forwardedPorts).forEach(unforward);
+  }
+
+  Future<ForwardedPort> _tryForward(int devicePort, int hostPort) async {
+    final List<String> interpolated = interpolateCommand(
+      _forwardPortCommand,
+      <String, String>{
+        'devicePort': '$devicePort',
+        'hostPort': '$hostPort'
+      },
+      additionalReplacementValues: _additionalReplacementValues
+    );
+
+    // launch the forwarding command
+    final Process process = await _processUtils.start(interpolated);
+
+    final Completer<ForwardedPort> completer = Completer<ForwardedPort>();
+
+    // read the outputs of the process, if we find a line that matches
+    // the configs forwardPortSuccessRegex, we complete with a successfully
+    // forwarded port
+    // Note that if that regex never matches, this will potentially run forever
+    // and the forwarding will never complete
+    final CustomDeviceLogReader reader = CustomDeviceLogReader(_deviceName)..listenToProcessOutput(process);
+    final StreamSubscription<String> logLinesSubscription = reader.logLines.listen((String line) {
+      if (_forwardPortSuccessRegex.hasMatch(line) && !completer.isCompleted) {
+        completer.complete(
+          ForwardedPort.withContext(hostPort, devicePort, process)
+        );
+      }
+    });
+
+    // if the process exits (even with exitCode == 0), that is considered
+    // a port forwarding failure and we complete with a null value.
+    unawaited(process.exitCode.whenComplete(() {
+      if (!completer.isCompleted) {
+        completer.complete(null);
+      }
+    }));
+
+    unawaited(completer.future.whenComplete(() {
+      logLinesSubscription.cancel();
+      reader.dispose();
+    }));
+
+    return completer.future;
+  }
+
+  @override
+  Future<int> forward(int devicePort, {int hostPort}) async {
+    int actualHostPort = (hostPort == 0 || hostPort == null) ? devicePort : hostPort;
+    int tries = 0;
+
+    while ((numTries == null) || (tries < numTries)) {
+      // when the desired host port is already forwarded by this Forwarder,
+      // choose another one
+      while (_forwardedPorts.any((ForwardedPort port) => port.hostPort == actualHostPort)) {
+        actualHostPort++;
+      }
+
+      final ForwardedPort port = await _tryForward(devicePort, actualHostPort);
+
+      if (port != null) {
+        _forwardedPorts.add(port);
+        return actualHostPort;
+      } else {
+        // null value means the forwarding failed (for whatever reason)
+        // increase port by one and try again
+        actualHostPort++;
+        tries++;
+      }
+    }
+
+    throw ToolExit('Forwarding port for custom device $_deviceName failed after $tries tries.');
+  }
+
+  @override
+  List<ForwardedPort> get forwardedPorts => List<ForwardedPort>.unmodifiable(_forwardedPorts);
+
+  @override
+  Future<void> unforward(ForwardedPort forwardedPort) async {
```

Comment:
Even though this method has no async now some could be added in the future (context for comment above on await unforwaard)

Reply:
Which method do you mean? the `forwardedPorts` property or `unforward`? Because `unforward` has an `async` right now

Btw seems like the documentation for `DevicePortForwarder.forwardedPorts` is wrong, it says it returns a Future of all the forwarded ports but it doesn't, just returns synchronously. (And the file name has a typo, the file is called `device_port_forwader.dart`)

Reply:
From L125

Reply:
but that has an `async` too, doesn't it?

Reply:
oops on the typo (FYI @jmagman )

At any rate, its important to await all of the futures you are using

Reply:
🤦‍♀️  https://github.com/flutter/flutter/pull/79481

Reply:
ah, yeah I've implemented that; should all be awaited now